Issue #1313 added artifacts filter support for files and Windows regi…

jnettesheim opened the pull request:
## One line description of pull request

Adding forensic artifact based filters for files and Windows registry keys- issue #1313.


## Description:
I've added a first pass at artifacts based filters.  This will only get files and registry keys for now.  I try to log warnings to catch instances where users are including other filters that aren't fully supported yet.

I'm storing the filter find_specs in the knowledge base, if you would prefer it not be there I can parse the file when needed, it would just lead to me parsing the file multiple times in some cases.

This probably also needs an end to end test and would be happy to get that in place.

Sincere apologies for the large PR.


**Related issue (if applicable):** adds to fixes for #1313

## Notes:
One thing to please look at:
LIne 267 in winreg.py: win_registry._registry_files.clear() ---- I had to call this to clear file not being closed errors, but I don't think we want that.

Also, appreciate any notes on glob expansion as I currently have it.

All contributions to Plaso undergo [code
review](https://github.com/log2timeline/plaso/wiki/Codereview). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://github.com/log2timeline/plaso/wiki/Style-guide).

One of the maintainers will examine your code, and may request changes.

## Checklist:
  - [Y ] Local tests pass
  - [?] Tests on TravisCI pass
  - [Y] Codacy passes (or flags issues that you think are acceptable) -- Used CodeFactor
  - [?] Coveralls indicates test coverage is sufficient

If [new dependencies](https://github.com/log2timeline/plaso/wiki/Adding-a-new-dependency) are required:
 - [ ] l2tdevtools has been updated

joachimmetz commented on plaso/cli/log2timeline_tool.py:
```diff
@@ -459,7 +466,18 @@ def ExtractEventsFromSources(self):
 
     filter_find_specs = None
-    if configuration.filter_file:
+    if configuration.artifacts_filter_file:
```
same remark as before, maybe create a helper method for this code and what to do about both filter file and artifacts filter file been set at the same time

jnettesheim replied:
Moved this to the tools.py so that it can be shared by image_export, log2timeline and psteal.

joachimmetz commented on plaso/cli/psteal_tool.py:
```diff
@@ -316,7 +321,17 @@ def ExtractEventsFromSources(self):
 
     filter_find_specs = None
-    if configuration.filter_file:
+    if configuration.artifacts_filter_file:
```
repeat of earlier comment about helper method

jnettesheim replied:
Moved this to the tools.py so that it can be shared by image_export, log2timeline and psteal.

joachimmetz commented on plaso/cli/status_view.py:
```diff
@@ -385,14 +389,17 @@ def SetMode(self, mode):
     self._mode = mode
 
-  def SetSourceInformation(self, source_path, source_type, filter_file=None):
+  def SetSourceInformation(self, source_path, source_type,
```
I opt to keep the style consistent with the rest of the source, so please line break after (

jnettesheim replied:
Done

joachimmetz commented on plaso/containers/sessions.py:
```diff
@@ -19,6 +19,7 @@ class Session(interface.AttributeContainer):
     aborted (bool): True if the session was aborted.
     analysis_reports_counter (collections.Counter): number of analysis reports
         per analysis plugin.
+    artifacts_filter_file (str): path to a file containing artifact definitions.
```
path to a file containing names of artifact definitions that are used for filtering file system and Windows Registry key paths.

jnettesheim replied:
Done

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+# -*- coding: utf-8 -*-
+"""Filter file."""
```
Artifacts filter file

jnettesheim replied:
Done

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+
+
+ARTIFACTS_FILTER_FILE = 'ARTIFACTS_FILTER_FILE'
+COMPATIBLE_DFWINREG_KEYS = [r'HKEY_LOCAL_MACHINE',
```
for code consistency please line break after [, replace r' by regular ' (escape \)

jnettesheim replied:
Done

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+      for source in definition.sources:
+        if source.type_indicator == artifact_types.TYPE_INDICATOR_FILE:
+          for path_entry in source.paths:
+            self.BuildFindSpecsFromFileArtifact(path_entry, source.separator,
```
please line breaks after ( for code consistency

jnettesheim replied:
Done

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+    except (KeyError, artifacts_errors.FormatError) as exception:
+      raise errors.BadConfigOption((
+          'Unable to read artifact definitions from: {0:s} with '
+          'error: ''{1!s}').format(self._path, exception))
```
why the double ' before {1!s} ?

jnettesheim replied:
Typo that pycharm completed for me, removed.

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+                                                      find_specs)
+        elif (source.type_indicator ==
+              artifact_types.TYPE_INDICATOR_WINDOWS_REGISTRY_VALUE):
+          # TODO: Handle Registry Values Once Supported in dfwinreg.
```
not sure what you mean here

jnettesheim replied:
A TODO to follow-up and see if there are plans for dfwinreg to support FindSpecs that specify a key-value pair, instead of the just the key path.  Right now I convert all key-value pairs to key paths only to build the FindSpecs.

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+        except KeyError as exception:
+          logging.error((
+              'Unable to expand path filter: {0:s} with error: '
+              '{1:s}').format(path, exception))
```
use {1!s} to 'cast' exception to a string

jnettesheim replied:
Done

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+    for path in self._ExpandGlobs(path_entry):
+      if path_attributes:
+        try:
+          if '%%environ_' in path:
```
make this more specific only known %%...%% should be replaced, also don't we do this already elsewhere in the codebase? maybe refactor to reuse existing functionatlity

jnettesheim replied:
I looked through trying to find where we might doing this already, but couldn't find it.  Can you point me towards it or should I implement?

jnettesheim replied:
I think I found and modified the path_helper to assist with this.  There's some more work that could be done to parse more variable like structures such as users.localappdata.

joachimmetz commented on plaso/engine/artifacts_filter_file.py:
```diff
@@ -0,0 +1,272 @@
+          environment variables.
+
+    Returns:
+      path_attributes dict[str]:  Dictionary containing the path attributes, per
```
path_attributes dict[str] => dict[str]

also 1 space after `:`

jnettesheim replied:
Done

joachimmetz commented on plaso/parsers/winreg.py:
```diff
@@ -170,23 +198,25 @@ def _ParseRecurseKeys(self, parser_mediator, root_key):
         break
 
-      matching_plugin = None
+      self._ParseKey(parser_mediator, registry_key)
```
what is the value of replacing this method with another method?

jnettesheim replied:
I build _ParseKey to avoid duplication of the same code in _ParseRecurseKeys and _ParseKeysFromFindSepcs.

joachimmetz replied:
ack

joachimmetz commented on plaso/parsers/winreg.py:
```diff
@@ -195,6 +225,7 @@ def ParseFileObject(self, parser_mediator, file_object, **kwargs):
       parser_mediator (ParserMediator): parser mediator.
       file_object (dfvfs.FileIO): a file-like object.
     """
+
```
no need for this white line

jnettesheim replied:
Done

joachimmetz commented on plaso/parsers/winreg.py:
```diff
@@ -212,10 +243,35 @@ def ParseFileObject(self, parser_mediator, file_object, **kwargs):
+    find_specs = parser_mediator.knowledge_base.GetValue(
+        artifacts_filter_file.ARTIFACTS_FILTER_FILE)
+
+    if (find_specs and
```
flatten this if else by changing the order in which things are checked

jnettesheim replied:
refactored this a bit

joachimmetz commented on tests/parsers/winreg.py:
```diff
@@ -102,6 +105,78 @@ def testParseSystem(self):
+    with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+      test_filter_file = artifacts_filter_file.ArtifactsFilterFile(
+          temp_file.name, knowledge_base)
+      temp_file.write(b'name: TestRegistryKey\n')
```
move the data to write to a string or list and write this in a single call, also looks like you can move the creation of the test data out of the with

jnettesheim replied:
Done

jnettesheim commented on plaso/cli/helpers/artifacts_filter_file.py:
```diff
@@ -0,0 +1,95 @@
+        '--artifacts_filter_file', '--artifacts-filter-file',
+        dest='artifacts_filter_file', type=str, default=None,
+        action='store', help=(
+            'Path to a directory containing artifact filter definitions, which '
```
Done

jnettesheim replied:
Done

jnettesheim commented on plaso/cli/helpers/artifacts_filter_file.py:
```diff
@@ -0,0 +1,95 @@
+      raise errors.BadConfigObject(
+          'Configuration object is not an instance of CLITool')
+
+    artifacts_filter_file = cls._ParseStringOption(options,
```
Done

jnettesheim commented on plaso/cli/image_export_tool.py:
```diff
@@ -415,7 +433,9 @@ def _ParseFilterOptions(self, options):
       raise errors.BadConfigOption(exception)
 
-    if self._filter_file:
+    if self._artifacts_filter_file:
```
Done

jnettesheim commented on plaso/cli/image_export_tool.py:
```diff
@@ -314,9 +320,21 @@ def _ExtractWithFilter(
-      filter_file_object = filter_file.FilterFile(filter_file_path)
-      find_specs = filter_file_object.BuildFindSpecs(
-          environment_variables=environment_variables)
+      if artifacts_filter_file_path:
```
Moved this to the tools.py so that it can be shared by image_export, log2timeline and psteal.

jnettesheim commented on plaso/cli/image_export_tool.py:
```diff
@@ -314,9 +320,21 @@ def _ExtractWithFilter(
+        find_specs = self._knowledge_base.GetValue(
+            artifacts_filter_file.ARTIFACTS_FILTER_FILE)[
+                artifact_types.TYPE_INDICATOR_FILE]
+      elif filter_file_path:
```
Moved this to the tools.py so that it can be shared by image_export, log2timeline and psteal.

jnettesheim commented on docs/plaso.engine.rst:
```diff
@@ -1,4 +1,5 @@
 plaso\.engine package
+plaso\.engine package
```
Don't know how this changed, but I think I got it back to the original.

jnettesheim commented:
Just one outstanding after the initial round now.

jnettesheim commented:
I've refactored this to now use the artifacts registry that is built as part of the pre-existing artifacts_definitions work.  I've also fixed a few other comments.  This will now accept artifacts as follows:

-  CLI entry of Common OS (Windows,Darwin,Linux) -> will run all matching this OS
-  Comma separate list of artifact names on CLI
-  A custom yaml file can also be uploaded and its names referenced

This probably needs a bit more polish, but wanted you to check in on the approach if possible.

joachimmetz commented:
please rebase from upstream master:
```
git fetch upstream && git pull --rebase upstream master
```

joachimmetz commented:
@jnettesheim please rebase from upstream master:

```
git fetch upstream && git pull --rebase upstream master
```

jnettesheim commented:
Rebased it, not sure if there is a better way to do it so that I didn't suddenly have all of the commits in my PR.

jnettesheim commented:
I used that command with that upstream, but fro my own fork of the repo .... Not sure if that impacted it.  I think it's probably easiest to just create a new PR now as it isn't obvious to Adam or myself how to fix this.  Does that work for you?

joachimmetz commented:
> Not sure if that impacted it.

Likely you'll need to do this on your feature branch that contains your PR. @jnettesheim please do not create a new fork try this:

```
git add remote upstream https://github.com/log2timeline/plaso.git
git fetch upstream && git pull --rebase upstream master
```

adamjnichols commented:
I've run a manual rebase on my own clone of j's fork, and there are several dozen conflicts.  I'm not certain the cleanest way out of this, but given the size of this PR and the delta between masters, it won't be easy no matter how it's done.

joachimmetz commented:
let me have a look

joachimmetz commented:
Done this:

```
git add remote upstream https://github.com/log2timeline/plaso.git
git checkout artifact-filters-feature
git checkout -b temp

git checkout artifact-filters-feature
git reset --hard 609212d9fe1129a15124c07a22208c32d776008b 
git fetch upstream && git pull --rebase upstream master
git pull --squash . temp

git rm -f docs/plaso.engine.rst

git commit --author "James <james.nettesheim@gmail.com>" -a -m "Added artifacts filter support for files and Windows Registry #1313"

git push -f
```

adamjnichols commented:
Yeah, no resets for me.  Not my code. :P

joachimmetz commented:
> Yeah, no resets for me. Not my code. :P

it was that or resolving merge conflicts manually, the squash just reapplies the changes ontop of the rebased branch

joachimmetz commented on plaso/cli/helpers/artifact_definitions.py:
```diff
@@ -86,11 +95,19 @@ def ParseOptions(cls, options, configuration_object):
       raise errors.BadConfigOption(
           'Unable to determine path to artifact definitions.')
 
+    custom_artifacts_path = getattr(
```
github does not allow me to create comment on 122 but you'll need to set custom_artifacts_path in the configuration object

jnettesheim replied:
I don't think I need to as I'm processing the file into the artifacts_registry within this helper?

joachimmetz replied:
Ack, I see you just load the artifact definitions from the path

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+          argparse group.
+    """
+    argument_group.add_argument(
+        '--artifact_filters', '--artifact-filters',
```
maybe split this in 2 different options, artifact_filters and artifact_filters_file?

jnettesheim replied:
Done.

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+            'formats. (1) Directly on the command line (comma separated), in a'
+            'in a file with one artifact name per line, or one operating system'
+            'specific keyword which will process all artifacts supporting that'
+            'OS (windows, linux, darwin).  Forensic artifacts are stored '
```
nit: Linux, MacOS (Darwin), Windows are names, so please start them with a captial

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+            'specific keyword which will process all artifacts supporting that'
+            'OS (windows, linux, darwin).  Forensic artifacts are stored '
+            'in .yaml files that are directly pulled from the artifact '
+            'definitions project. You can also specify a custom artifacts yaml'
```
nit: plaso does not use the double spaces after the end of a sentence. Please use a single space.

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+            'definitions project. You can also specify a custom artifacts yaml'
+            'file (see --custom_artifact_definitions).  Artifact definitions '
+            'can be used to describe and quickly collect data of interest, such'
+            ' as specific files or Windows Registry keys.'))
```
nit: for consistency put the space at the end of the previous line (inside the string)

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+      BadConfigObject: when the configuration object is of the wrong type.
+      BadConfigOption: if the required artifact definitions are not defined.
+    """
+
```
- white line

jnettesheim replied:
Done

joachimmetz replied:
I meant remove white line ( `-` gets translated to a list :| )

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+        '--artifact_filters', '--artifact-filters',
+        dest='artifact_filters', type=str, default=None,
+        action='store', help=(
+            'Names of forensic artifact definitions, provided in the following'
```
spaces missing at the end of line (inside the string)

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+      raise errors.BadConfigObject(
+          'Configuration object is not an instance of CLITool')
+
+    artifact_filters = cls._ParseStringOption(
```
the `_ParseStringOption` method can return None, please check its result first before applying lower()

also chaining command make troubleshooting harder, my recommendation is to limit the number of chained commands

https://github.com/log2timeline/plaso/blob/c084c2db262457fba1bc353832e1c110fd57628b/plaso/cli/helpers/interface.py#L53

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+    elif artifact_filters:
+      artifact_filters = artifact_filters.split(',')
+
+    setattr(configuration_object, '_artifact_filters',
```
nit: this looks like it can fit on a single line

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+
+    if artifact_filters and os.path.isfile(artifact_filters):
+      with open(artifact_filters) as f:
+        artifact_filters = f.read().splitlines()
```
please don't chain commands

joachimmetz commented on plaso/cli/helpers/artifact_filters.py:
```diff
@@ -0,0 +1,77 @@
+        options, 'artifact_filters').lower()
+
+    if artifact_filters and os.path.isfile(artifact_filters):
+      with open(artifact_filters) as f:
```
Please don't use abbreviations, it makes the code hard to read. Use file_object instead of f

Also see: https://github.com/log2timeline/plaso/wiki/Style-guide

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/image_export_tool.py:
```diff
@@ -8,6 +8,7 @@
 import os
 import textwrap
 
+
```
- white line

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/image_export_tool.py:
```diff
@@ -296,8 +299,12 @@ def _ExtractWithFilter(
+          listed comma separated.
       filter_file_path (str): path of the file that contains the filter
-          expressions.
+          expressions or artifact definitions.
```
I opt: "or names of artifact definitions."

also why overload filter_file_path ?

jnettesheim replied:
Done.

joachimmetz commented on plaso/cli/image_export_tool.py:
```diff
@@ -296,8 +299,12 @@ def _ExtractWithFilter(
       source_path_specs (list[dfvfs.PathSpec]): path specifications to extract.
       destination_path (str): path where the extracted files should be stored.
       output_writer (CLIOutputWriter): output writer.
+      artifacts_registry (ArtifactRegistry): Artifacts registry object.
```
```
      artifacts_registry (artifacts.ArtifactDefinitionsRegistry]): artifact
          definitions registry.
```

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/image_export_tool.py:
```diff
@@ -296,8 +299,12 @@ def _ExtractWithFilter(
       output_writer (CLIOutputWriter): output writer.
+      artifacts_registry (ArtifactRegistry): Artifacts registry object.
+      artifact_filters_path (str): path of the file that contains the
+          names of the artifacts filter definitions or definitions directly
```
Let's not overload paths here, this makes the code hard to follow and more error prone.

Split artifact_filters_path and artifact_filters

joachimmetz commented on plaso/cli/image_export_tool.py:
```diff
@@ -313,10 +320,8 @@ def _ExtractWithFilter(
-      filter_file_object = filter_file.FilterFile(filter_file_path)
-      find_specs = filter_file_object.BuildFindSpecs(
-          environment_variables=environment_variables)
+      find_specs = tools.FindSpecsGetter().GetFindSpecs(artifacts_registry,
```
It is unclear what `FindSpecsGetter` means, please rename to something that indicates the action of the method, maybe `CreateFilterFindSpecs` or `GetFilterFindSpecs`?

Also please do not chain commands, this makes the code harder to debug.

jnettesheim replied:
Done.

joachimmetz commented on plaso/cli/log2timeline_tool.py:
```diff
@@ -458,13 +462,9 @@ def ExtractEventsFromSources(self):
-      filter_file_object = filter_file.FilterFile(configuration.filter_file)
-      filter_find_specs = filter_file_object.BuildFindSpecs(
-          environment_variables=environment_variables)
+    filter_find_specs = tools.FindSpecsGetter().GetFindSpecs(
```
same comment as before

jnettesheim replied:
Done.

joachimmetz commented on plaso/cli/psteal_tool.py:
```diff
@@ -315,13 +318,9 @@ def ExtractEventsFromSources(self):
-      filter_file_object = filter_file.FilterFile(configuration.filter_file)
-      filter_find_specs = filter_file_object.BuildFindSpecs(
-          environment_variables=environment_variables)
+    filter_find_specs = tools.FindSpecsGetter().GetFindSpecs(
```
same comment as before

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/status_view.py:
```diff
@@ -317,6 +318,9 @@ def PrintExtractionStatusHeader(self, processing_status):
         'Source type\t: {0:s}\n'.format(self._source_type))
 
+    if self._artifact_filters:
+      self._output_writer.Write('Artifact filters \t: {0:s}\n'.format(
```
no need for space after `filters`:
'Artifact filters \t: {0:s}\n

either:
'Artifact filters\t\t: {0:s}\n
'Artifact filters\t: {0:s}\n

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/status_view.py:
```diff
@@ -389,14 +393,18 @@ def SetMode(self, mode):
 
-  def SetSourceInformation(self, source_path, source_type, filter_file=None):
+  def SetSourceInformation(
+      self, source_path, source_type, artifact_filters=None,
```
lines 397 and 398 look like they can fit on a single line

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/status_view.py:
```diff
@@ -389,14 +393,18 @@ def SetMode(self, mode):
     Args:
       source_path (str): path of the source.
       source_type (str): source type.
+      artifact_filters (Optional[str]): artifact filters.
```
please reword, maybe something like: names of the artifact definitions to use as filter.

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/tools.py:
```diff
@@ -507,3 +511,38 @@ def Write(self, string):
+  def GetFindSpecs(
+      self, artifacts_registry, artifact_filter_names, filter_file_path,
+      knowledge_base):
+    """Get Find Specs from artifacts or filter file if available.
```
Get => Retrieves or Determines ?

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/tools.py:
```diff
@@ -507,3 +511,38 @@ def Write(self, string):
       sys.stdout.write(string)
+
+
+class FindSpecsGetter(object):
```
several nits:
* please don't overload the term getter (also see: https://en.wikipedia.org/wiki/Mutator_method)
* I opt we make this a method part of BaseEngine instead?

joachimmetz commented on plaso/cli/tools.py:
```diff
@@ -507,3 +511,38 @@ def Write(self, string):
+       knowledge_base (KnowledgeBase): Knowledge base.
+    """
+    if artifact_filter_names and filter_file_path:
+      raise RuntimeError('Please only specify one type of filter file, '
```
consistency nit: in the plaso codebase we typically break after `(` and continue with 4 space continuation indentation

jnettesheim replied:
Done

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+
+
+ARTIFACT_FILTERS = 'ARTIFACT_FILTERS'
+COMPATIBLE_DFWINREG_KEYS = ['HKEY_LOCAL_MACHINE',
```
break after `[`

also should this be a frozenset?

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+from plaso.engine import path_helper
+
+
+ARTIFACT_FILTERS = 'ARTIFACT_FILTERS'
```
Please make these class constants instead of module constants

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+class ArtifactFilters(object):
+  """Forensic artifacts filters.
+
+  Build extraction and parsing filters from forensic artifact definitions.
```
I opt to remove  "and parsing" seeing at this point we only support to create path filters

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+# -*- coding: utf-8 -*-
+"""Artifact filters processing."""
```
"Artifact filters processing" => Helper to create filters based on artifact definitions.

jnettesheim replied:
Done

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+RECURSIVE_GLOB_LIMIT = 10
+
+
+class ArtifactFilters(object):
```
ArtifactDefinitionsFilterHelper (or equiv)

rationale:
* this is not a Filter as in the sense of other filters,
* technically this class is processing artifact definitions not the artifacts themselves

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+
+
+class ArtifactFilters(object):
+  """Forensic artifacts filters.
```
=> Helper to create filters based on artifact definitions.

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+
+  Build extraction and parsing filters from forensic artifact definitions.
+
+  Forensic artifacts are defined in:
```
maybe: For more information about Forensic Artifacts see:

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+  """
+
+  def __init__(self, artifacts_registry, filters, knowledge_base):
+    """Initializes a filter file.
```
"filter file" => "artifact definitions filter helper"

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+    """Initializes a filter file.
+
+    Args:
+      path (str): path to a file that contains one or more forensic artifacts.
```
artifacts_registry missing from docstring

      artifacts_registry (artifacts.ArtifactDefinitionsRegistry]): artifact
          definitions registry.

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+    self._knowledge_base = knowledge_base
+
+  def BuildFindSpecs(self, environment_variables=None):
+    """Build find specification from a forensic artifacts file.
```
"from a forensic artifacts file." => "from artifact definitions."

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+              artifact_types.TYPE_INDICATOR_WINDOWS_REGISTRY_VALUE):
+          # TODO: Handle Registry Values Once Supported in dfwinreg.
+          logging.warning(('Unable to handle Registry Value, extracting '
+                           'key only: {0:s} ').format(source.key_value_pairs))
```
source.key_value_pairs is a list of tuples, so {:s} is likely the wrong formatter

use an explicit join e.g. ','.join(['(key: {0:s}, value: {1:s})'.format(key_path, value_name) for key_path, value_name in source.key_value_pairs])

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+# -*- coding: utf-8 -*-
```
Maybe copy this to a separate PR and review this first?

This PR is getting big

jnettesheim replied:
Done

jnettesheim replied:
Merged in #1883

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -0,0 +1,239 @@
+      environment_variables (Optional[list[EnvironmentVariableArtifact]]):
+          environment variables.
+    """
+    find_specs = {}
```
I stopped my review here.

jnettesheim replied:
Merged in #1883

joachimmetz commented:
@jnettesheim let's chat about this PR, it is getting large I have a lot of nits ;) I opt we break out engine.ArtifactFilters into a separate PR first

jnettesheim commented:
@joachimmetz - I think this is pretty close now, PTAL.

joachimmetz commented:
Please make linter pass: https://travis-ci.org/log2timeline/plaso/jobs/396010009

joachimmetz commented on plaso/cli/extraction_tool.py:
```diff
@@ -75,6 +75,8 @@ def _CreateProcessingConfiguration(self, knowledge_base):
     # TODO: pass preferred_encoding.
     configuration = configurations.ProcessingConfiguration()
+    configuration.artifact_filters = self._artifact_filters
+    configuration.artifacts_registry = self._artifacts_registry
```
Re: artifacts_registry

Please don't pass complex objects via configuration configuration.artifacts_registry. Configuration is meant to pass settings. If you need pass complex objects do this via the methods.

jnettesheim replied:
It seems we were storing a complete ArtifactDefinitionsRegistry() in the configuration object, prior to these changes.  However, I don't think it was ever used, so I've refactored the helper to store the artifact definitions path and customer artifact definitions path only.  All tests are passing after refactoring, so I think I've accounted for any other impact.

joachimmetz commented on plaso/engine/configurations.py:
```diff
@@ -211,6 +212,8 @@ class ProcessingConfiguration(interface.AttributeContainer):
   def __init__(self):
     """Initializes a process configuration object."""
     super(ProcessingConfiguration, self).__init__()
+    self.artifacts_registry = None
```
Re: artifacts_registry

Please don't pass complex objects via configuration configuration.artifacts_registry. Configuration is meant to pass settings. If you need pass complex objects do this via the methods.

also it is not defined in the docstring.

jnettesheim replied:
It seems we were storing a complete ArtifactDefinitionsRegistry() in the configuration object, prior to these changes.  However, I don't think it was ever used, so I've refactored the helper to store the artifact definitions path and customer artifact definitions path only.  All tests are passing after refactoring, so I think I've accounted for any other impact.

joachimmetz commented on plaso/cli/helpers/artifact_definitions.py:
```diff
@@ -56,6 +65,7 @@ def ParseOptions(cls, options, configuration_object):
       BadConfigObject: when the configuration object is of the wrong type.
       BadConfigOption: if the required artifact definitions are not defined.
     """
+
```
remove white line (`-` white line)

jnettesheim replied:
Done.

joachimmetz commented on plaso/cli/helpers/artifact_definitions.py:
```diff
@@ -86,11 +96,20 @@ def ParseOptions(cls, options, configuration_object):
+        registry.ReadFromFile(reader, custom_artifacts_path)
+      elif custom_artifacts_path and not os.path.isfile(custom_artifacts_path):
+        raise errors.BadConfigOption(
+            'No such artifacts filter file: {0:s}.'.format(
```
nit: looks like this and the next line could fit on a single line

jnettesheim replied:
Unfortunately, they do not fit on a single line.

joachimmetz commented on plaso/cli/image_export_tool.py:
```diff
@@ -297,8 +299,12 @@ def _ExtractWithFilter(
       output_writer (CLIOutputWriter): output writer.
-      filter_file_path (str): path of the file that contains the filter
-          expressions.
+      artifacts_registry (artifacts.ArtifactDefinitionsRegistry]): artifact
```
remove the `]` in `(artifacts.ArtifactDefinitionsRegistry])`

jnettesheim replied:
Done

joachimmetz commented on plaso/cli/image_export_tool.py:
```diff
@@ -297,8 +299,12 @@ def _ExtractWithFilter(
-          expressions.
+      artifacts_registry (artifacts.ArtifactDefinitionsRegistry]): artifact
+          definitions registry.
+      artifact_filters (list[str]): Names of artifact definitions that are
```
Names  => names (use lower case for consistency)

jnettesheim replied:
Done.

joachimmetz commented on plaso/containers/sessions.py:
```diff
@@ -184,6 +189,8 @@ class SessionStart(interface.AttributeContainer):
   """Session start attribute container.
 
   Attributes:
+    artifact_filters (list[str]): Names of artifact definitions that are
```
Names => names (lower case)

jnettesheim replied:
Done.

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -20,23 +20,29 @@ class ArtifactDefinitionsFilterHelper(object):
       artifacts_registry (artifacts.ArtifactDefinitionsRegistry]): artifact
           definitions registry.
-      artifact_definitions (list[str]): artifact definition names to filter.
+      artifact_filters (list[str]): Names of artifact definitions that are
```
Names => names (lower case)

jnettesheim replied:
Done.

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -20,23 +20,29 @@ class ArtifactDefinitionsFilterHelper(object):
+  KNOWLEDGE_BASE_VALUE = 'ARTIFACT_FILTERS'
 
-  _COMPATIBLE_REGISTRY_KEY_PATH_PREFIXES = ['HKEY_LOCAL_MACHINE']
+  COMPATIBLE_REGISTRY_KEY_PATH_PREFIXES = frozenset([
```
Why change these? Are these intended to be used outside the class (except by unit tests)?

jnettesheim replied:
I use them in winreg.py as well, but just refactored this a bit into a function so reverting this.

joachimmetz commented on plaso/engine/artifact_filters.py:
```diff
@@ -108,7 +114,7 @@ def BuildFindSpecs(self, environment_variables=None):
           # BuildFindSpecsFromRegistryArtifact.
           for key_path in set([
-              key_path for key_path, _ in source.key_value_pairs]):
+              key_value['key'] for key_value in source.key_value_pairs]):
```
Why change this?

jnettesheim replied:
The way we had re-worked it led to it building a set that only contained "key" in it.  For example, with:  {key: 'HKEY_LOCAL_MACHINE\System\ControlSet002\Control\Session Manager', value: 'BootExecute'}, 

key_path for key_path, _ in source.key_value_pairs]): only extracts the literal "key", whereas we want the value of "key" -- HKEY_LOCAL_MACHINE\System\ControlSet002....

joachimmetz commented on plaso/engine/engine.py:
```diff
@@ -272,3 +278,42 @@ def SupportsGuppyMemoryProfiling(cls):
+    """Build Find Specs from artifacts or filter file if available.
+
+    Args:
+       knowledge_base_object (KnowledgeBase): Knowledge base.
```
Consistency nit start with lower case after `:`

jnettesheim replied:
Done.

joachimmetz commented on plaso/engine/engine.py:
```diff
@@ -272,3 +278,42 @@ def SupportsGuppyMemoryProfiling(cls):
+    Raises:
+      RuntimeError: if no valid FindSpecs are built.
+    """
+
```
remove white line

jnettesheim replied:
Done.

joachimmetz commented on plaso/engine/engine.py:
```diff
@@ -272,3 +278,42 @@ def SupportsGuppyMemoryProfiling(cls):
+          environment_variables=environment_variables)
+
+    if (artifact_filter_names or filter_file_path) and not find_specs:
+      raise RuntimeError('Error processing filters, no valid specifications'
```
I opt to break after `(` for consistency

jnettesheim replied:
Done.

joachimmetz commented on tests/engine/artifact_filters.py:
```diff
@@ -81,7 +81,7 @@ def testBuildFindSpecsWithFileSystem(self):
         environment_variables=[environment_variable])
     find_specs_per_source_type = knowledge_base.GetValue(
-        test_filter_file._KNOWLEDGE_BASE_VALUE)
+        test_filter_file.KNOWLEDGE_BASE_VALUE)
```
just override pylint check for access to protected members in the tests

jnettesheim replied:
I do access this in winreg.py as well

joachimmetz commented on tests/parsers/winreg.py:
```diff
@@ -102,6 +107,66 @@ def testParseSystem(self):
+    find_specs = {
+        test_filter_file.KNOWLEDGE_BASE_VALUE : knowledge_base.GetValue(
+            test_filter_file.KNOWLEDGE_BASE_VALUE)}
+    storage_writer = self._ParseFile(['SYSTEM'], parser,
```
please break after `(`

jnettesheim replied:
Done.

joachimmetz commented on plaso/parsers/winreg.py:
```diff
@@ -169,23 +196,25 @@ def _ParseRecurseKeys(self, parser_mediator, root_key):
+      find_specs (dfwinreg.FindSpecs): Keys to search for.
+    """
+    searcher = dfwinreg_registry_searcher.WinRegistrySearcher(win_registry)
+    for registry_key_path in list(searcher.Find(find_specs=find_specs)):
```
could you use iter() instead of list() ?

jnettesheim replied:
Done.

joachimmetz commented on plaso/parsers/winreg.py:
```diff
@@ -205,16 +234,43 @@ def ParseFileObject(self, parser_mediator, file_object, **kwargs):
+      try:
+        self._ParseRecurseKeys(parser_mediator, root_key)
+      except IOError as exception:
+        parser_mediator.ProduceExtractionError('{0:s}'.format(exception))
```
add 1 white line below

jnettesheim replied:
Done.

joachimmetz commented:
Mainly nits remaining. If you address them and the linter issues, this CL could be good to go.

joachimmetz commented:
@Onager for awareness override of CodeFactor results (it indicated minor code duplication primarily)
